sycl/doc/design/VirtualFunctions.md

```diff
@@ -0,0 +1,428 @@
+runtime to link them together. Device images which depend on optional kernel
+features are linked only if those features are supported by a target device and
+dummy versions of those device images are used otherwise.
+
+This way we can emit single unified version of LLVM IR where vtables reference
+all device virtual functions, but their definitions are outlined and linked
+back dynamically based on device capabilities.
+
+For AOT flow, we don't do outlining and dynamic linking, but instead do direct
+cleanup of virtual functions which are incompatible with a target device.
+
+## Design
+
+### Changes to the SYCL header files
+
+New compile-time properties `indirectly_callable` and `calls_indirectly` should
+be implemented in accordance with the corresponding [design document][2]:
+
+- `indirectly_callable` property should lead to emission of
+  `"indirectly-callable"="set"` function attribute, where "set" is a string
+  representation of the property template parameter.
+- `calls_indirectly` property should lead to emission of
+  `"calls-indirectly"="set1,set2"`, where "set1" and "set2" are string
+  representations of the property template parameters.
+
+In order to convert a type to a string, [\__builtin_sycl_unique_stable_name][3]
+could be used.
+
+**TODO**: `calls_indirectly` requires compile-time concatenation of strings.
+Document how it should be done.
+
+`indirectly_callable` property is applied to functions using "custom" (comparing
+to other properties) `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro. This
+is done to allow implementations to attach some extra attributes alongside the
+property. In particular, functions marked with the macro should be considered
+SYCL device functions and compiler should emit diagnostics if those functions
+do not conform with the SYCL 2020 specification. To achieve that and avoid
+extending FE to parse strings within properties, the aforementioned macro should
+also set `sycl_device` attribute:
+
+```
+#define SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY(SetId)                    \
+  __attribute__((sycl_device)) [[__sycl_detail__::add_ir_attribute_function(   \
+      "indirectly-callable", __builtin_sycl_unique_stable_name(SetId))]]
+```
+
+### Changes to the compiler front-end
+
+Most of the handling for virtual functions happens in middle-end and thanks to
+compile-time properties, no extra work is required to propagate necessary
+information down to passes from headers.
+
+However, we do need to filter out those virtual functions which are not
+considered to be device  as defined by the [extension specification][1], such
+as:
+
+- virtual member functions annotated with `indirectly_callable` compile-time
+  property should be emitted into device code;
+- virtual member function *not* annotated with `indirectly_callable`
+  compile-time property should *not* be emitted into device code;
+
+There is no need to actually check which exact property is applied to a
```

Comment:
`SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` should result in `__attribute__((sycl_device)) `right? Am I correct in understanding that therefore all virtual functions which are required to be emitted will have `__attribute__((sycl_device)) ` and those which shouldn't be emitted won't? In that case there is no need to check for `add_ir_attribute_function`.

Reply:
> `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` should result in `__attribute__((sycl_device)) `right?

Right

> Am I correct in understanding that therefore all virtual functions which are required to be emitted will have `__attribute__((sycl_device)) ` and those which shouldn't be emitted won't? In that case there is no need to check for `add_ir_attribute_function`.

Yes, that's a good point. Device virtual functions are expected to be directly marked as device ones, so we could just use that existing attribute, I will update the document

Reply:
Reworded this section in 8c8185205f8c4c6e8e08eb055c94c27bb936c40d